lib/Backend/Lower.cpp

```diff
@@ -17931,6 +17934,70 @@ Lowerer::GenerateFastInlineArrayPop(IR::Instr * instr)
     GenerateHelperToArrayPopFastPath(instr, doneLabel, bailOutLabelHelper);
 }
 
+void
+Lowerer::GenerateFastInlineHasOwnProperty(IR::Instr * instr)
+{
+    Assert(instr->m_opcode == Js::OpCode::CallDirect);
+
+    //CallDirect src2
+    IR::Opnd * linkOpnd = instr->GetSrc2();
+    //ArgOut_A_InlineSpecialized
+    IR::Instr * tmpInstr = linkOpnd->AsSymOpnd()->m_sym->AsStackSym()->m_instrDef;
+
+    IR::Opnd * argsOpnd[2] = { 0 };
+    bool result = instr->FetchOperands(argsOpnd, 2);
+    Assert(result);
+    AnalysisAssert(argsOpnd[0] && argsOpnd[1]);
+
+    // fast path case where hasOwnProperty is being called using a property name loaded via a for-in loop
+    if (!argsOpnd[1]->GetValueType().IsString()
+        || argsOpnd[0]->GetValueType().IsNotObject()
+        || !argsOpnd[0]->IsRegOpnd()
+        || !argsOpnd[1]->IsRegOpnd()
+        || !argsOpnd[1]->AsRegOpnd()->m_sym->m_isSingleDef
+        || (argsOpnd[1]->AsRegOpnd()->m_sym->m_instrDef->m_opcode != Js::OpCode::BrOnEmpty
+            && argsOpnd[1]->AsRegOpnd()->m_sym->m_instrDef->m_opcode != Js::OpCode::BrOnNotEmpty))
+    {
+        return;
+    }
+
+    IR::LabelInstr *doneLabel = InsertLabel(false, instr->m_next);
+    IR::LabelInstr *labelHelper = InsertLabel(true, instr);
+    IR::Instr * insertInstr = labelHelper;
+    IR::RegOpnd * thisObj = argsOpnd[0]->AsRegOpnd();
+    IR::Opnd * forInEnumeratorOpnd = argsOpnd[1]->AsRegOpnd()->m_sym->m_instrDef->GetSrc1();
+
+    // go to helper if we can't use JIT fastpath
+    IR::Opnd * canUseJitFastPathOpnd = GetForInEnumeratorFieldOpnd(forInEnumeratorOpnd, Js::ForInObjectEnumerator::GetOffsetOfCanUseJitFastPath(), TyInt8);
```

Comment:
Sure. I ended up going more conservative than originally coded to be safe and win seemed about same so I stuck with this variant. Wonder though if maybe it would be better to go more aggressive and check only that we have initial type, and that type is lockable.

I still see a bit of time left (3% under HasOwnProperty), but I think remaining cases are mostly when using different object than you are enumerating against.